collab-folder/src/core/favorites.rs

```diff
@@ -0,0 +1,109 @@
+use crate::core::FavoritesInfo;
+use anyhow::bail;
+use collab::preclude::{lib0Any, Array, ArrayRefWrapper, ReadTxn, TransactionMut, Value, YrsValue};
+
+use serde::{Deserialize, Serialize};
+
+pub struct FavoritesArray {
+  container: ArrayRefWrapper,
+}
+
+impl FavoritesArray {
+  pub fn new(root: ArrayRefWrapper) -> Self {
+    Self { container: root }
+  }
+  ///Gets all favorite views in form of FavoriteRecord[]
+  pub fn get_all_favorites(&self) -> Vec<FavoritesInfo> {
+    let txn = self.container.transact();
+    let items = self.get_all_favorites_with_txn(&txn);
+    items
+      .into_iter()
+      .map(|item| FavoritesInfo { id: item.id })
+      .collect::<Vec<_>>()
+  }
+  pub fn get_all_favorites_with_txn<T: ReadTxn>(&self, txn: &T) -> Vec<FavoriteRecord> {
+    let mut favorites = vec![];
+    for value in self.container.iter(txn) {
+      if let YrsValue::Any(any) = value {
+        favorites.push(FavoriteRecord::from(any))
+      }
+    }
+    favorites
+  }
+
+  /// Deletes a favorited record to be used when a view / views is / are unfavorited
+  pub fn delete_favorites<T: AsRef<str>>(&self, ids: Vec<T>) {
+    self.container.with_transact_mut(|txn| {
+      self.delete_favorites_with_txn(txn, ids);
+    })
+  }
+
+  pub fn delete_favorites_with_txn<T: AsRef<str>>(&self, txn: &mut TransactionMut, ids: Vec<T>) {
+    for id in &ids {
+      if let Some(pos) = self
+        .get_all_favorites_with_txn(txn)
+        .into_iter()
+        .position(|item| item.id == id.as_ref())
+      {
+        self.container.remove_with_txn(txn, pos as u32);
+      }
+    }
+  }
+
+  /// Adds a favorited record to be used when a view / views is / are favorited
+  pub fn add_favorites(&self, favorite_records: Vec<FavoriteRecord>) {
+    self.container.with_transact_mut(|txn| {
+      self.add_favorites_with_txn(txn, favorite_records);
+    })
+  }
+
+  pub fn add_favorites_with_txn(
+    &self,
+    txn: &mut TransactionMut,
+    favorite_records: Vec<FavoriteRecord>,
+  ) {
+    for favorite_record in favorite_records {
+      self.container.push_back(txn, favorite_record);
+    }
+  }
+
+  pub fn clear(&self) {
+    self.container.with_transact_mut(|txn| {
+      let len = self.container.iter(txn).count();
+      self.container.remove_range(txn, 0, len as u32);
+    });
+  }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct FavoriteRecord {
+  pub id: String,
+  #[serde(default)]
+  pub workspace_id: String,
```

Comment:
If you use #[serde(default)] here, it implies that the workspace_id could be an empty string, which is not allowed. The workspace_id must not be an empty string.